Feature elasticsearch

lauritssn opened the pull request:
Add Elasticsearch as a database, vector database and cache file storage. Initial version - needs a little more testing to ensure everything works as expected - especially on the retrieval part. Some files may need to be removed before merge.

lauritssn commented:
You can use this sign off: Signed-off-by: Laurits Søgaard Nielsen <lsn@werktoej.dk>

kichanyurd commented:
@lauritssn we can't modify your commit messages ourselves, they have to by you - unfortunately.

Fortunately - it's easy to fix. You can ask Claude/GPT for help with doing this but the process is basically:

1. Setup sign off using the rules under CONTRIBUTING.md in git
2. git rebase interactively to squash your commits into one, and finally amend it to include the sign off note

lauritssn committed:
feat: Add Elasticsearch as database, vector database and cache storage

- Add ElasticsearchDocumentDatabase for document storage
- Add ElasticsearchVectorDatabase for vector similarity search
- Add SDK options to configure elasticsearch stores
- Add configuration documentation and examples
- Add comprehensive test coverage

Signed-off-by: Laurits Søgaard Nielsen <lsn@werktoej.dk>

lauritssn committed:
feat: Add cursor-based pagination support to Elasticsearch adapter

- Add SortDirection enum and Cursor dataclass to persistence/common.py
- Add encode_cursor/decode_cursor utilities in app_modules/common.py
- Implement pagination in ElasticsearchDocumentCollection.find()
- Update capabilities.py to use FindResult interface
- Add pagination tests for Elasticsearch adapter

Signed-off-by: Laurits Søgaard Nielsen <lsn@werktoej.dk>

lauritssn committed:
fix: Improve test isolation and fix API compatibility in Elasticsearch tests

- Fix delete_collection pattern to include _vecdb_ in index name matching
- Add unique test prefixes per test to prevent index conflicts
- Update tests for NullEmbedder rename (was NoOpEmbedder)
- Fix test_document_upgrade to create proper index mappings before inserting
- Add VectorDatabase type annotation and import to SDK

Signed-off-by: Laurits Søgaard Nielsen <lsn@werktoej.dk>

lauritssn committed:
fix: Handle association document versions 0.6.0 and 0.7.0 in guidelines store

The association schema hasn't changed since 0.5.0, but the document
loader needs to handle the newer version strings.

Signed-off-by: Laurits Søgaard Nielsen <lsn@werktoej.dk>

lauritssn committed:
docs: Move Elasticsearch documentation to correct locations

- Move elasticsearch_document_database.md to docs/adapters/persistence/
- Move elasticsearch_configuration.md to docs/adapters/vector_db/
- Update headers to match existing adapter documentation style
- Remove broken internal reference link

Signed-off-by: Laurits Søgaard Nielsen <lsn@werktoej.dk>

mc-dorzo commented on src/parlant/core/guidelines.py:
```diff
@@ -376,7 +376,8 @@ async def _association_document_loader(
 
-        if doc["version"] == "0.5.0":
+        # 0.5.0+ all use the same schema - no migration needed
+        if doc["version"] in ("0.5.0", "0.6.0", "0.7.0"):
```
Good observation of this bug! But actually, this is not a clean fix.
We want to update every document to the latest version. For 0.5.0 and 0.6.0, we want to update the document to 0.7.0.

lauritssn replied:
ok - should be fixed now

mc-dorzo commented on src/parlant/sdk.py:
```diff
@@ -3114,9 +3121,17 @@ def __init__(
-        session_store: Literal["transient", "local"] | str | SessionStore = "transient",
-        customer_store: Literal["transient", "local"] | str | CustomerStore = "transient",
-        variable_store: Literal["transient", "local"] | str | ContextVariableStore = "transient",
+        session_store: Literal["transient", "local", "elasticsearch"]
```
Since very few users are using Elasticsearch for persistence, we don't want to include it in the SDK. 
With MongoDB, even though we didn't include it in the options, we implemented it, so you need to set an environment variable to enable it.
What we did with the other adapters is add an explanation to the Elasticsearch.md file on how to enable this adapter using configure_container. 
Please take a look at the docs/adapters/persistence/snowflake.md for an example.

lauritssn committed:
Remove elasticsearch from SDK persistence store literals

- Remove 'elasticsearch' from session_store, customer_store, variable_store,
  and cache_store Literal types in SDK Server constructor
- Keep 'elasticsearch' as a valid option for vector_store
- Update persistence elasticsearch.md with SDK/Module Setup section showing
  how to use configure_container for Elasticsearch persistence (like Snowflake)
- Update vector_db elasticsearch.md with Quick Start showing both direct SDK
  usage (vector_store='elasticsearch') and advanced configure_container approach
- Add new SDK exports: AgentStore, SessionDocumentStore, CustomerDocumentStore,
  ContextVariableDocumentStore, IdGenerator, EventEmitterFactory, VectorDatabase
- Update installation examples to use uv instead of pip

Signed-off-by: Laurits Søgaard Nielsen <lsn@werktoej.dk>

lauritssn committed:
fix: Migrate guideline tag associations to latest version 0.7.0

Update _association_document_loader to migrate all older document versions
(0.3.0, 0.4.0, 0.5.0, 0.6.0) to the latest version 0.7.0 instead of just
passing them through or migrating to an intermediate version.

Signed-off-by: Laurits Søgaard Nielsen <lsn@werktoej.dk>

lauritssn committed:
fix: Remove elasticsearch from SDK vector_store option

Remove the remaining "elasticsearch" option from vector_store Literal type
and its implementation code. Users who want to use Elasticsearch for vector
storage should use configure_container to set it up, similar to other
adapters like Snowflake and Qdrant.

This keeps the SDK unchanged from its original state before the Elasticsearch
PR, as requested by the maintainers.

Signed-off-by: Laurits Søgaard Nielsen <lsn@werktoej.dk>

lauritssn committed:
fix: Revert all SDK changes to match upstream develop

The SDK should remain unchanged. Users who want to use Elasticsearch
should use configure_container to set it up, as documented in
docs/adapters/persistence/elasticsearch.md and
docs/adapters/vector_db/elasticsearch.md.

Signed-off-by: Laurits Søgaard Nielsen <lsn@werktoej.dk>

lauritssn commented:
@mc-dorzo Can you please check if everything is as expected now?
